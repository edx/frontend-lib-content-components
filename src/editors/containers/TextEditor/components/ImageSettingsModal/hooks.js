import React from 'react';

import { StrictDict } from '../../../../utils';
import * as module from './hooks';

// Simple wrappers for useState to allow easy mocking for tests.
export const state = {
  dimensions: (val) => React.useState(val),
  locked: (val) => React.useState(val),
  local: (val) => React.useState(val),
  lockInitialized: (val) => React.useState(val),
  altText: (val) => React.useState(val),
  isDecorative: (val) => React.useState(val),
};

export const dimKeys = StrictDict({
  height: 'height',
  width: 'width',
});

/**
 * findGcd(numerator, denominator)
 * Find the greatest common denominator of a ratio or fraction.
 * @param {number} numerator - ratio numerator
 * @param {number} denominator - ratio denominator
 * @return {number} - ratio greatest common denominator
 */
export const findGcd = (a, b) => (b ? findGcd(b, a % b) : a);
const checkEqual = (d1, d2) => (d1.height === d2.height && d1.width === d2.width);

/**
 * getValidDimensions({ dimensions, local, locked })
 * Find valid ending dimensions based on start state, request, and lock state
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after move ({ height, width })
 */
export const getValidDimensions = ({
  dimensions,
  local,
  locked,
}) => {
  const out = {};
  let iter;
  const { minInc } = locked;
  const isMin = dimensions.height === minInc.height;

  const keys = (local.height !== dimensions.height)
    ? { changed: dimKeys.height, other: dimKeys.width }
    : { changed: dimKeys.width, other: dimKeys.height };

  const direction = local[keys.changed] > dimensions[keys.changed] ? 1 : -1;

  // don't move down if already at minimum size
  if (direction < 0 && isMin) { return dimensions; }
  // find closest valid iteration of the changed field
  iter = Math.max(Math.round(local[keys.changed] / minInc[keys.changed]), 1);
  // if closest valid iteration is current iteration, move one iteration in the change direction
  if (iter === (dimensions[keys.changed] / minInc[keys.changed])) { iter += direction; }

  out[keys.changed] = iter * minInc[keys.changed];
  out[keys.other] = out[keys.changed] * (locked[keys.other] / locked[keys.changed]);

  return out;
};

/**
 * newDimensions({ dimensions, local, locked })
 * Returns the local dimensions if unlocked or unchanged, and otherwise returns new valid
 * dimensions.
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after attempted move ({ height, width })
 */
export const newDimensions = ({ dimensions, local, locked }) => (
  (!locked || checkEqual(local, dimensions))
    ? local
    : module.getValidDimensions({ dimensions, local, locked })
);

/**
 * lockDimensions({ dimensions, lockInitialized, setLocked })
 * Lock dimensions if lock initialized.  Store minimum valid increment on lock so
 * that we don't have re-compute.
 * @param {obj} dimensions - current stored dimensions
 * @param {bool} lockInitialized - has the lock state initialized?
 * @param {func} setLocked - set lock state
 */
export const lockDimensions = ({ dimensions, lockInitialized, setLocked }) => {
  if (!lockInitialized) { return; }

  // find minimum viable increment
  let gcd = findGcd(dimensions.width, dimensions.height);
  if ([dimensions.width, dimensions.height].some(v => !Number.isInteger(v / gcd))) {
    gcd = 1;
  }
  const minInc = { width: dimensions.width / gcd, height: dimensions.height / gcd, gcd };
  setLocked({ ...dimensions, minInc });
};

/**
 * dimensionLockHooks({ dimensions })
 * Returns a set of hooks pertaining to the dimension locks.
 * Locks the dimensions initially, on lock initialization.
 * @param {obj} dimensions - current stored dimensions
 * @return {obj} - dimension lock hooks
 *   {func} initializeLock - enable the lock mechanism
 *   {obj} locked - current locked state
 *   {func} lock - lock the current dimensions
 *   {func} unlock - unlock the dimensions
 */
export const dimensionLockHooks = ({ dimensions }) => {
  const [locked, setLocked] = module.state.locked(null);
  const [lockInitialized, setLockInitialized] = module.state.lockInitialized(null);
  const lock = () => module.lockDimensions({ lockInitialized, dimensions, setLocked });

  React.useEffect(lock, [lockInitialized]);

  return {
    initializeLock: () => setLockInitialized(true),
    locked,
    lock,
    unlock: () => setLocked(null),
  };
};

/**
 * dimensionHooks()
 * Returns an object of dimension-focused react hooks.
 * @return {obj} - dimension hooks
 *   {func} onImgLoad - initializes image dimension fields
 *     @param {object} selection - selected image object with possible override dimensions.
 *     @return {callback} - image load event callback that loads dimensions.
 *   {object} locked - current locked state
 *   {func} lock - lock current dimensions
 *   {func} unlock - unlock dimensions
 *   {object} value - current dimension values
 *   {func} setHeight - set height
 *     @param {string} - new height string
 *   {func} setWidth - set width
 *     @param {string} - new width string
 *   {func} updateDimensions - set dimensions based on state
 */
export const dimensionHooks = () => {
  const [dimensions, setDimensions] = module.state.dimensions(null);
  const [local, setLocal] = module.state.local(null);
  const setAll = (value) => {
    setDimensions(value);
    setLocal(value);
  };
  const {
    initializeLock,
    lock,
    locked,
    unlock,
  } = module.dimensionLockHooks({ dimensions });
  return {
    onImgLoad: (selection) => ({ target: img }) => {
      setAll({
        height: selection.height || img.naturalHeight,
        width: selection.width || img.naturalWidth,
      });
      initializeLock();
    },
    locked,
    lock,
    unlock,
    value: local,
    setHeight: (height) => setLocal({ ...local, height: parseInt(height, 10) }),
    setWidth: (width) => setLocal({ ...local, width: parseInt(width, 10) }),
    updateDimensions: () => setAll(module.newDimensions({ dimensions, local, locked })),
  };
};

/**
 * altTextHooks(savedText)
 * Returns a set of react hooks focused around alt text
 * @return {obj} - alt text hooks
 *   {string} value - alt text value
 *   {func} setValue - set alt test value
 *     @param {string} - new alt text
 *   {bool} isDecorative - is the image decorative?
 *   {func} setIsDecorative - set isDecorative field
 *     @param {bool} isDecorative
 */
export const altTextHooks = (savedText) => {
  const [value, setValue] = module.state.altText(savedText || '');
  const [isDecorative, setIsDecorative] = module.state.isDecorative(false);
  return {
    value,
    setValue,
    isDecorative,
    setIsDecorative,
  };
};

/**
 * onInputChange(handleValue)
 * Simple event handler forwarding the event target value to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target value.
 */
export const onInputChange = (handleValue) => (e) => handleValue(e.target.value);

/**
 * onCheckboxChange(handleValue)
 * Simple event handler forwarding the event target checked prop to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target checked prop.
 */
export const onCheckboxChange = (handleValue) => (e) => handleValue(e.target.checked);

/**
 * onSave({ altText, dimensions, isDecorative, saveToEditor })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {object} dimension - image dimensions ({ width, height })
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} saveToEditor - save method for submitting image settings.
 */
export const onSaveClick = ({
  altText,
  dimensions,
  isDecorative,
  saveToEditor,
}) => () => saveToEditor({
  altText,
  dimensions,
  isDecorative,
});

/**
 * isSaveDisabled(altText)
 * Returns true the save button should be disabled (altText is missing and not decorative)
 * @param {object} altText - altText hook object
 *   {bool} isDecorative - is the image decorative?
 *   {string} value - alt text value
 * @return {bool} - should the save button be disabled?
 */
export const isSaveDisabled = (altText) => !altText.isDecorative && (altText.value === '');

export default {
  altText: altTextHooks,
  dimensions: dimensionHooks,
  isSaveDisabled,
  onCheckboxChange,
  onInputChange,
  onSaveClick,
};
