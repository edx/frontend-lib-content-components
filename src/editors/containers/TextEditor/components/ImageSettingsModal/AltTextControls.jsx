import React from 'react';
import PropTypes from 'prop-types';
import { Form } from '@edx/paragon';

import * as hooks from './hooks';

/**
 * Wrapper for alt-text input and isDecorative checkbox control
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} setIsDecorative - handle isDecorative change event
 * @param {func} setValue - update alt-text value
 * @param {string} value - current alt-text value
 */
export const AltTextControls = ({
  isDecorative,
  setIsDecorative,
  setValue,
  value,
}) => (
  <Form.Group className="mt-4.5">
    <Form.Label as="h4">Accessibility</Form.Label>
    <Form.Control
      className="mt-4.5"
      type="input"
      value={value}
      disabled={isDecorative}
      onChange={hooks.onInputChange(setValue)}
      floatingLabel="Alt Text"
    />
    <Form.Checkbox
      className="mt-4.5 decorative-control-label"
      checked={isDecorative}
      onChange={hooks.onCheckboxChange(setIsDecorative)}
    >
      <Form.Label>
        This image is decorative (no alt text required).
      </Form.Label>
    </Form.Checkbox>
  </Form.Group>
);
AltTextControls.propTypes = {
  isDecorative: PropTypes.bool.isRequired,
  value: PropTypes.string.isRequired,
  setValue: PropTypes.func.isRequired,
  setIsDecorative: PropTypes.func.isRequired,
};

export default AltTextControls;
