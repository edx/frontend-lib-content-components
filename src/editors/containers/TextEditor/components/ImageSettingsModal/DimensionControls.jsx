import React from 'react';
import PropTypes from 'prop-types';
import {
  Form,
  Icon,
  IconButton,
} from '@edx/paragon';
import {
  Locked,
  Unlocked,
} from '@edx/paragon/icons';

import hooks from './hooks';

/**
 * Wrapper for image dimension inputs and the lock checkbox.
 * @param {obj} locked - locked dimension object
 * @param {func} lock - lock dimensions
 * @param {func} setHeight - updates dimensions based on new height
 * @param {func} setWidth - updates dimensions based on new width
 * @param {func} unlock - unlock dimensions
 * @param {func} updateDimensions - update dimensions callback
 * @param {obj} value - local dimension values { height, width }
 */
export const DimensionControls = ({
  locked,
  lock,
  setHeight,
  setWidth,
  unlock,
  updateDimensions,
  value,
}) => ((value !== null) && (
  <Form.Group>
    <Form.Label as="h4">Image Dimensions</Form.Label>
    <div className="mt-4.5">
      <Form.Control
        className="dimension-input"
        type="number"
        value={value.width}
        min={1}
        onChange={hooks.onInputChange(setWidth)}
        onBlur={updateDimensions}
        floatingLabel="Width"
      />
      <Form.Control
        className="dimension-input"
        type="number"
        value={value.height}
        min={1}
        onChange={hooks.onInputChange(setHeight)}
        onBlur={updateDimensions}
        floatingLabel="Height"
      />
      <IconButton
        className="d-inline-block"
        alt={locked ? 'unlock dimensions' : 'lock dimensions'}
        iconAs={Icon}
        src={locked ? Locked : Unlocked}
        onClick={locked ? unlock : lock}
      />
    </div>
  </Form.Group>
));
DimensionControls.defaultProps = {
  locked: null,
  value: {
    height: 100,
    width: 100,
  },
};
DimensionControls.propTypes = ({
  value: PropTypes.shape({
    height: PropTypes.number,
    width: PropTypes.number,
  }),
  setHeight: PropTypes.func.isRequired,
  setWidth: PropTypes.func.isRequired,
  locked: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
  }),
  lock: PropTypes.func.isRequired,
  unlock: PropTypes.func.isRequired,
  updateDimensions: PropTypes.func.isRequired,
});

export default DimensionControls;
