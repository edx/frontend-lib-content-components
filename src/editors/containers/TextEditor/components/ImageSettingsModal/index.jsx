import React from 'react';
import PropTypes from 'prop-types';
import { Button, Image } from '@edx/paragon';
import { ArrowBackIos } from '@edx/paragon/icons';

import BaseModal from '../BaseModal';

import AltTextControls from './AltTextControls';
import DimensionControls from './DimensionControls';
import hooks from './hooks';
import './index.scss';

/**
 * Modal display wrapping the dimension and alt-text controls for image tags
 * inserted into the TextEditor TinyMCE context.
 * Provides a thumbnail and populates dimension and alt-text controls.
 * @param {bool} isOpen - is the modal open?
 * @param {func} close - close the modal
 * @param {obj} selection - current image selection object
 * @param {func} saveToEditor - save the current settings to the editor
 * @param {func} returnToSelection - return to image selection
 */
export const ImageSettingsModal = ({
  isOpen,
  close,
  selection,
  saveToEditor,
  returnToSelection,
}) => {
  const dimensions = hooks.dimensions();
  const altText = hooks.altText();
  const onSaveClick = hooks.onSaveClick({
    saveToEditor,
    dimensions: dimensions.value,
    altText: altText.value,
    isDecorative: altText.isDecorative,
  });
  return (
    <BaseModal
      title="Image Settings"
      close={close}
      isOpen={isOpen}
      confirmAction={(
        <Button
          variant="primary"
          onClick={onSaveClick}
          disabled={hooks.isSaveDisabled(altText)}
        >
          Save
        </Button>
      )}
    >
      <Button
        onClick={returnToSelection}
        variant="link"
        size="inline"
        iconBefore={ArrowBackIos}
      >
        Replace image
      </Button>
      <br />
      <div className="d-flex flex-row m-2 img-settings-form-container">
        <div className="img-settings-thumbnail-container">
          <Image
            className="img-settings-thumbnail"
            onLoad={dimensions.onImgLoad(selection)}
            src={selection.externalUrl}
          />
        </div>
        <hr className="h-100 bg-primary-200 m-0" />
        <div className="img-settings-form-controls">
          <DimensionControls {...dimensions} />
          <AltTextControls {...altText} />
        </div>
      </div>
    </BaseModal>
  );
};

ImageSettingsModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  selection: PropTypes.shape({
    url: PropTypes.string,
    externalUrl: PropTypes.string,
    altText: PropTypes.bool,
  }).isRequired,
  saveToEditor: PropTypes.func.isRequired,
  returnToSelection: PropTypes.func.isRequired,
};
export default ImageSettingsModal;
