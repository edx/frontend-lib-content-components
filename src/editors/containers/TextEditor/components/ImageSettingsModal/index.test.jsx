import React from 'react';
import { shallow } from 'enzyme';
import ImageSettingsModal from '.';

jest.mock('./AltTextControls', () => 'AltTextControls');
jest.mock('./DimensionControls', () => 'DimensionControls');

jest.mock('./hooks', () => ({
  dimensions: () => ({
    value: { width: 12, height: 13 },
    onImgLoad: jest.fn(
      (selection) => ({ 'hooks.dimensions.onImgLoad.callback': { selection } }),
    ).mockName('hooks.dimensions.onImgLoad'),
  }),
  altText: () => ({
    value: 'alternative Taxes',
    isDecorative: false,
  }),
  onSaveClick: (args) => ({ 'hooks.onSaveClick': args }),
  isSaveDisabled: (args) => ({ 'hooks.isSaveDisabled': args }),
}));

describe('ImageSettingsModal', () => {
  const props = {
    isOpen: false,
    selection: { selected: 'image data' },
  };
  beforeEach(() => {
    props.close = jest.fn().mockName('props.close');
    props.saveToEditor = jest.fn().mockName('props.saveToEditor');
    props.returnToSelection = jest.fn().mockName('props.returnToSelector');
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(shallow(<ImageSettingsModal {...props} />)).toMatchSnapshot();
    });
  });
});
