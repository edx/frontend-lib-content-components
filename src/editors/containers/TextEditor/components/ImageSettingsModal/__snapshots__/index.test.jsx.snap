// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSettingsModal render snapshot 1`] = `
<BaseModal
  close={[MockFunction props.close]}
  confirmAction={
    <Button
      disabled={
        Object {
          "hooks.isSaveDisabled": Object {
            "isDecorative": false,
            "value": "alternative Taxes",
          },
        }
      }
      onClick={
        Object {
          "hooks.onSaveClick": Object {
            "altText": "alternative Taxes",
            "dimensions": Object {
              "height": 13,
              "width": 12,
            },
            "isDecorative": false,
            "saveToEditor": [MockFunction props.saveToEditor],
          },
        }
      }
      variant="primary"
    >
      Save
    </Button>
  }
  isOpen={false}
  title="Image Settings"
>
  <Button
    onClick={[MockFunction props.returnToSelector]}
    size="inline"
    variant="link"
  >
    Replace image
  </Button>
  <br />
  <div
    className="d-flex flex-row m-2 img-settings-form-container"
  >
    <div
      className="img-settings-thumbnail-container"
    >
      <Image
        className="img-settings-thumbnail"
        onLoad={
          Object {
            "hooks.dimensions.onImgLoad.callback": Object {
              "selection": Object {
                "selected": "image data",
              },
            },
          }
        }
      />
    </div>
    <hr
      className="h-100 bg-primary-200 m-0"
    />
    <div
      className="img-settings-form-controls"
    >
      <DimensionControls
        onImgLoad={
          [MockFunction hooks.dimensions.onImgLoad] {
            "calls": Array [
              Array [
                Object {
                  "selected": "image data",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Object {
                  "hooks.dimensions.onImgLoad.callback": Object {
                    "selection": Object {
                      "selected": "image data",
                    },
                  },
                },
              },
            ],
          }
        }
        value={
          Object {
            "height": 13,
            "width": 12,
          }
        }
      />
      <AltTextControls
        isDecorative={false}
        value="alternative Taxes"
      />
    </div>
  </div>
</BaseModal>
`;
